File: src/java/foundgine-core/src/main/java/com/foundgine/core/semantic/SemanticType.java
package com.foundgine.core.semantic;

import java.math.BigDecimal;
import java.time.temporal.TemporalAccessor;
import java.util.Collection;
import java.util.Objects;
import java.util.UUID;

/** Provider-independent semantic type information. */
public sealed interface SemanticType permits SemanticType.Scalar, SemanticType.EnumType, SemanticType.ObjectType, SemanticType.CollectionType {
    record Scalar(SemanticScalarKind kind) implements SemanticType { public Scalar { Objects.requireNonNull(kind); } }
    record EnumType(String name) implements SemanticType { public EnumType { Objects.requireNonNull(name); } }
    record ObjectType(String name) implements SemanticType { public ObjectType { Objects.requireNonNull(name); } }
    record CollectionType(SemanticType elementType) implements SemanticType { public CollectionType { Objects.requireNonNull(elementType); } }

    static SemanticType fromClass(Class<?> type) {
        Objects.requireNonNull(type);
        if (type == byte[].class) return new Scalar(SemanticScalarKind.BYTES);
        if (type.isArray()) return new CollectionType(fromClass(type.getComponentType()));
        if (type.isEnum()) return new EnumType(type.getSimpleName());
        if (type == String.class || type == Character.class || type == char.class) return new Scalar(SemanticScalarKind.STRING);
        if (type == boolean.class || type == Boolean.class) return new Scalar(SemanticScalarKind.BOOLEAN);
        if (type == byte.class || type == Byte.class || type == short.class || type == Short.class || type == int.class || type == Integer.class)
            return new Scalar(SemanticScalarKind.INT32);
        if (type == long.class || type == Long.class) return new Scalar(SemanticScalarKind.INT64);
        if (type == float.class || type == Float.class || type == double.class || type == Double.class || type == BigDecimal.class)
            return new Scalar(SemanticScalarKind.DECIMAL);
        if (TemporalAccessor.class.isAssignableFrom(type) || java.util.Date.class.isAssignableFrom(type)) return new Scalar(SemanticScalarKind.DATETIME);
        if (type == UUID.class) return new Scalar(SemanticScalarKind.GUID);
        if (Collection.class.isAssignableFrom(type)) return new CollectionType(new ObjectType("Object"));
        return new ObjectType(type.getSimpleName());
    }
}

File: src/csharp/Foundgine.Core/Semantic/SemanticType.cs
namespace Foundgine.Core.Semantic;

/// <summary>
/// Provider-independent semantic type information. CLR types may still be
/// retained by adapters/providers, but semantic consumers should use this
/// contract rather than depending on System.Type.
/// </summary>
public abstract record SemanticType
{
    public sealed record Scalar(SemanticScalarKind Kind) : SemanticType;

    public sealed record Enum(string Name) : SemanticType;

    public sealed record Object(string Name) : SemanticType;

    public sealed record Collection(SemanticType ElementType) : SemanticType;

    public static SemanticType FromClrType(Type type)
    {
        ArgumentNullException.ThrowIfNull(type);

        var nullable = Nullable.GetUnderlyingType(type);
        if (nullable is not null)
            return FromClrType(nullable);

        if (type == typeof(byte[])) return new Scalar(SemanticScalarKind.Bytes);
        if (type.IsArray)
            return new Collection(FromClrType(type.GetElementType()!));

        if (type.IsEnum)
            return new Enum(type.Name);

        if (type == typeof(string) || type == typeof(char)) return new Scalar(SemanticScalarKind.String);
        if (type == typeof(bool)) return new Scalar(SemanticScalarKind.Boolean);
        if (type == typeof(byte) || type == typeof(sbyte) || type == typeof(short) || type == typeof(ushort) ||
            type == typeof(int)) return new Scalar(SemanticScalarKind.Int32);
        if (type == typeof(uint) || type == typeof(long) || type == typeof(ulong))
            return new Scalar(SemanticScalarKind.Int64);
        if (type == typeof(float) || type == typeof(double) || type == typeof(decimal))
            return new Scalar(SemanticScalarKind.Decimal);
        if (type == typeof(DateTime) || type == typeof(DateTimeOffset)) return new Scalar(SemanticScalarKind.DateTime);
        if (type == typeof(Guid)) return new Scalar(SemanticScalarKind.Guid);

        var enumerable = type.GetInterfaces()
            .Append(type)
            .FirstOrDefault(x => x.IsGenericType && x.GetGenericTypeDefinition() == typeof(IEnumerable<>));
        if (enumerable is not null)
            return new Collection(FromClrType(enumerable.GetGenericArguments()[0]));

        return new Object(type.Name);
    }
}

public enum SemanticScalarKind : byte
{
    String,
    Int32,
    Int64,
    Decimal,
    Boolean,
    DateTime,
    Guid,
    Bytes
}